pkg/controller/instance/instance_controller_test.go

```diff
@@ -0,0 +1,117 @@
+// +build integration
+
+package instance
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/onsi/ginkgo"
+
+	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1alpha1"
+	"github.com/kudobuilder/kudo/pkg/util/kudo"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/onsi/gomega"
+
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const timeout = time.Second * 5
+const tick = time.Millisecond * 500
+
+func TestRestartController(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	gomega.RegisterFailHandler(ginkgo.Fail)
+
+	stopMgr, mgrStopped, c := startTestManager(g)
+
+	log.Printf("Given an existing instance 'foo-instance' and operator 'foo-operator'")
+	in := &v1alpha1.Instance{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo-instance", Namespace: "default", Labels: map[string]string{kudo.OperatorLabel: "foo-operator"}},
+		Spec: v1alpha1.InstanceSpec{
+			OperatorVersion: v1.ObjectReference{
+				Name:      "foo-operator",
+				Namespace: "default",
+			},
+			Parameters: map[string]string{"param": "value"},
+		},
+	}
+	instanceKey, _ := client.ObjectKeyFromObject(in)
+	assert.NoError(t, c.Create(context.TODO(), in))
+	defer c.Delete(context.TODO(), in)
+
+	ov := &v1alpha1.OperatorVersion{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo-operator", Namespace: "default"},
+		TypeMeta:   metav1.TypeMeta{Kind: "OperatorVersion", APIVersion: "kudo.dev/v1alpha1"},
+		Spec: v1alpha1.OperatorVersionSpec{
+			Plans: map[string]v1alpha1.Plan{"deploy": {}, "update": {}},
+			Parameters: []v1alpha1.Parameter{
+				{
+					Name:    "param",
+					Default: kudo.String("default"),
+				},
+			},
+		},
+	}
+	assert.NoError(t, c.Create(context.TODO(), ov))
+	defer c.Delete(context.TODO(), ov)
+
+	log.Print("And a deploy plan that was already run")
+	assert.Eventually(t, func() bool { return instancePlanFinished(instanceKey, "deploy", c) }, timeout, tick)
+
+	log.Print("When we stop the manager")
+	close(stopMgr)
+	mgrStopped.Wait()
+
+	log.Print("And update the instance parameter value")
+	err := c.Get(context.TODO(), instanceKey, in)
```

Comment:
Is this paranoid programming or do we really expect the instance to disappear? 😄

Reply:
haha this is actually necessary because the instance changed in the meantime (controller updated some fields, so we need to pull it again). I'll add comment :)